Feature/migrate fbw to reactquery

selankon opened the pull request:
Migrate to React Query library instead of use the old redux system

selankon committed:
Implement useCreateNetwork

selankon committed:
chore(fbw): fix bad api call

selankon committed:
chore(fbw): refactor searchNetworks to react query

selankon committed:
chore(fbw): refactor setNetwork to react query

selankon committed:
chore(fbw): store expected host and network state

selankon committed:
fix(fbw): join network select not filling

selankon committed:
refactor(fbw): delete remaining redux code

selankon committed:
fix(fbw): bad import

selankon committed:
refactor(fbw): file names to naming convention

selankon committed:
chore(fbw): delete API mock

germanferrero commented on plugins/lime-plugin-fbw/src/FbwApi.js:
```diff
@@ -1,13 +1,13 @@
-export const setNetwork = (api, { file, hostname }) =>
-	api.call('lime-fbw', 'set_network', { file, hostname });
+export const setNetwork = ({ file, hostname, network }) =>
+	api.call('lime-fbw', 'set_network', { file, hostname, network });
```
The lime-fbw set_network endpoint doesn't expect a network attribute in the body. Is this something intentional?

selankon replied:
Well, when `setNetwork` is called, on the original code, it pass a param called `network`. See: [Scan.js](https://github.com/libremesh/lime-app/blob/develop/plugins/lime-plugin-fbw/src/containers/Scan.js#L48)

```javascript
if (state.apname && isValidHostname(state.hostname, true)) {
			setNetwork({
				file: state.file,
				hostname: state.hostname,
				network: state.community
			});
		}
```

germanferrero replied:
Ah, ok! The backend is not using it anyway, we can remove it :)

germanferrero commented on plugins/lime-plugin-fbw/src/containers/Scan.js:
```diff
@@ -3,26 +3,49 @@ import { useState, useEffect } from 'preact/hooks';
 import Toast from 'components/toast';
 import { isValidHostname, slugify } from 'utils/isValidHostname';
-import { showNotification } from '../../../../src/store/actions';
+// import { showNotification } from '../../../../src/store/actions';
```
commented code to be deleted

germanferrero commented on plugins/lime-plugin-fbw/src/FbwApi.js:
```diff
@@ -1,13 +1,13 @@
 import api from 'utils/uhttpd.service';
-import { from } from 'rxjs';
+// import { from } from 'rxjs';
```
commented code to be deleted

germanferrero commented on plugins/lime-plugin-fbw/src/containers/Scan.js:
```diff
@@ -3,26 +3,49 @@ import { useState, useEffect } from 'preact/hooks';
+			toggleForm('setting')();
+		},
+		onError: () => {
+			setStatus('error')
```
You can avoid using:
```
const [status, setStatus] = useState()
```
And take it directly from:
```const [setNetwork, { isError: isSetNetworkError}] = useSetNetwork()```
and
```const [searchNetworks, { isError: isSeachNetworkError}] = useSearchNetworks()```

selankon replied:
And  on  `isSeachNetworkError` and `isSetNetworkError` I have to do `setStatus('error')`? 

Also, what about the `onSuccess` result? Its ok as is now?

germanferrero replied:
isSeachNetworkError and isSetNetworkError are booleans you can use directly in your conditional rendering.

onSuccess callback looks god!

germanferrero commented on plugins/lime-plugin-fbw/src/containers/Scan.js:
```diff
@@ -3,26 +3,49 @@ import { useState, useEffect } from 'preact/hooks';
+	const [status, setStatus] = useState()
+
+	const [searchNetworks, { isLoading: isSubmitting}] = useSearchNetworks({
+		onSuccess: (payload) => {
```
Hmm, the previous code doesn't help... But it would make much more sense to have a query called useGetNetworks to do the "GET action" and the mutation called useSearchNetworks to do the "POST action".

Then you would use something like this to obtain the networks.
const { data: networks, isLoading }  = useGetNetworks();
(this query can right now call search_networks endpoint with rescan:false parameter).

And useSearchNetworks would do call setQueryData on the useGetNetworks on success. That would make the component re-render and update `networks`.

In this way, you also avoid using useState for something that it's backend state, rather than frontend state.

germanferrero commented on plugins/lime-plugin-fbw/src/containers/Scan.js:
```diff
@@ -148,20 +172,14 @@ export const Scan = ({ searchNetworks, setNetwork, toggleForm, status, networks
-});
-
-export default connect(mapStateToProps, mapDispatchToProps)(Scan);
+const getApName = ({ ap = '', file = '' }) => {
```
It's okey to do this hear, but just to give you options:
This data enrichment could also be done at query level.
You can check the "useAssocList" query to checkout what I mean.

germanferrero reviewed (commented):
@selankon great work! I've add some comments you may want to address :)

selankon committed:
chore(fbw): delete comented code

selankon committed:
refactor(fbw): try to implement useGetNetworks query

selankon committed:
chore(fbw): fix bad setQueryData invocation

selankon committed:
chore(fbw): implement useSearchNetworks to modify useGetNetworks query

Fixing also:

https://github.com/libremesh/lime-app/pull/334#discussion_r809318442
https://github.com/libremesh/lime-app/pull/334#discussion_r809327161

selankon committed:
fix(fbw): unused parameter

See https://github.com/libremesh/lime-app/pull/334/files/e5f27d0f55a2bb9b196089f17b4d9c7e6df77862#r809291076

selankon committed:
chore(fbw): implement better error status management

Fixing:

https://github.com/libremesh/lime-app/pull/334/files/e5f27d0f55a2bb9b196089f17b4d9c7e6df77862#r809304350

selankon committed:
chore(fbw): merge seGetNetworks query implementation

This come from https://github.com/libremesh/lime-app/pull/334#discussion_r809318442

germanferrero committed:
chore(fbw): tweaks to enable testing

germanferrero committed:
test(fbw): add test for checking step.

selankon committed:
Merge pull request #1 from germanferrero/migrate-fbw-test-expected-network

Migrate fbw: test expectedHost/expectedNetwork handling

selankon committed:
chore(fbw): implement setExpectedHost/Network after creation

selankon committed:
test(fbw): add missing "."

Expected string ends in "."

germanferrero reviewed (approved):
Great work, thanks @selankon !! :tada: :smile:

selankon committed:
Merge branch 'develop' into feature/migrate-fbw-to-reactquery

selankon committed:
chore(fbw): do eslint fixes

selankon commented:
Hi @germanferrero I merged from develop and applied eslint rules but I faced that there are some files not involving to FBW that need lintern revisions. Cold we step forward with this PR? 

```
➜ npx eslint .

limeapp/lime-app/plugins/lime-plugin-changeNode/src/changeNodePage.js
  62:41  error  Missing "key" prop for element in iterator  react/jsx-key

limeapp/lime-app/plugins/lime-plugin-fbw/src/FbwPage.spec.js
  61:9  error  Expect must have a corresponding matcher call  jest/valid-expect
  68:9  error  Expect must have a corresponding matcher call  jest/valid-expect

limeapp/lime-app/plugins/lime-plugin-metrics/src/metricsPage.js
  154:4  error  Missing "key" prop for element in iterator  react/jsx-key

limeapp/lime-app/plugins/lime-plugin-pirania/src/screens/voucherList.js
  68:7  error  Missing "key" prop for element in iterator  react/jsx-key

limeapp/lime-app/plugins/lime-plugin-rx/src/rxPage.js
  126:8  error  Missing "key" prop for element in iterator  react/jsx-key

limeapp/lime-app/src/components/app.js
  35:5  error  Missing "key" prop for element in iterator  react/jsx-key
  43:5  error  Missing "key" prop for element in iterator  react/jsx-key
  53:5  error  Missing "key" prop for element in iterator  react/jsx-key
  64:5  error  Missing "key" prop for element in iterator  react/jsx-key

limeapp/lime-app/src/components/tabs/index.js
  7:5  error  Missing "key" prop for element in iterator  react/jsx-key

limeapp/lime-app/src/utils/app.context.js
  3:1  error  'preact' import is duplicated  no-duplicate-imports

✖ 12 problems (12 errors, 0 warnings)
```
